### PR TITLE
Adding SAS URL Lifetime as a parameter, so this can be customised

### LINKED ISSUE #4224: Allow custom SAS URL lifetime in Azure builder
In the Azure builder, allow custom lifetime for SAS URL, this is currently hard coded to one month

I need SAS URLs with a lifetime longer than one month, so currently am using the Azure CLI once I have built an image to create another longer lived SAS URL, would be good to have this configurable within Packer

### PR DESCRIPTION
**Pull request by elliottsam**
At present the SAS URL that comes from the Azure builder is hard coded to only one month, this change makes this setting configurable in the json template file

Closes #4224

### CODE REVIEW

File: builder/azure/arm/config.go [hunk 1]
```diff
@@ -109,6 +109,9 @@ type Config struct {
 
 	Comm communicator.Config `mapstructure:",squash"`
 	ctx  *interpolate.Context
+
+	// SAS URL lifetime in months
+	SasUrlLifetime int `mapstructure:"sas_url_lifetime"`
```
Comment (boumenot) [inline builder/azure/arm/config.go]: This should be a duration.  It gives much better granularity, and is more intuitive.

**Comment by elliottsam** [comment]
Okay, I have made those changes, how's it looking now?

File: builder/azure/arm/config.go [hunk 2]
```diff
@@ -362,6 +368,23 @@ func setCustomData(c *Config) error {
 	return nil
 }
 
+func setSasUrlDuration(c *Config) error {
+	if c.SasUrlLifetime == "" {
+		c.SasUrlLifetime = "720h"
```
Comment (boumenot) [inline builder/azure/arm/config.go]: I would express this as 30 * 24 * time.Hour to make the value a little clearer.

File: builder/azure/arm/config.go [hunk 3]
```diff
@@ -362,6 +368,23 @@ func setCustomData(c *Config) error {
...
+	}
+	match := regexp.MustCompile("^([0-9]+[hm])([0-9]+[m])?$")
+	if match.MatchString(c.SasUrlLifetime) != true {
+		err := fmt.Errorf("Failed to convert SasUtlLifetime to duration, format is invalid: %s", c.SasUrlLifetime)
+		return err
+	}
+	b, err := time.ParseDuration(c.SasUrlLifetime)
```
Comment (boumenot) [inline builder/azure/arm/config.go]: Why do you need to match a regular expression, and use ParseDuration.  ParseDuration is sufficient, and will be more robust.

File: builder/azure/arm/config_test.go [hunk 4]
```diff
@@ -737,6 +737,51 @@ func TestConfigShouldRejectMissingCustomDataFile(t *testing.T) {
 	}
 }
 
+func TestConfigShouldRejectMalformedSasUrlLifetimeDuration(t *testing.T) {
```
Comment (boumenot) [inline builder/azure/arm/config_test.go]: Excellent!

File: website/source/docs/builders/azure.html.md [hunk 5]
```diff
@@ -109,6 +109,9 @@ builder.
 
     CLI example `azure vm sizes -l westus`
 
+-   `sas_url_lifetime` (string) Specify the lifetime of the SAS URLs returned, if this is not set it will default to 30
+    days, this should take the format of a duration of hours and minutes, i.e. 720h or 2h30m
```
Comment (boumenot) [inline website/source/docs/builders/azure.html.md]: NIT: this should take -> this takes

File: builder/azure/arm/config.go (hunk 3 above)
Comment (elliottsam) [inline builder/azure/arm/config.go]: Sorry fairly new to Go and not used time.Duration before, so was unsure what was needed, will sort this out

**Comment by elliottsam** [comment]
Sorted those last couple of bits, should I squash these commits into one?

File: builder/azure/arm/config.go (hunk 3 above)
Comment (boumenot) [inline builder/azure/arm/config.go]: No need to apologize!  We love any and all contributions.

**Comment by boumenot** [comment]
I remove an oldcomment, and I think I broke the build because the code needs to be reformatted.  I thought I was going to make an easy change, and then just commit the PR.  I guess not.

```text
==> Checking that code complies with gofmt requirements...
gofmt needs running on the following files:
./builder/azure/arm/config.go
You can use the command: `make fmt` to reformat code.
```

Would you please reformat, squash, and I'll commit the PR.

**Comment by mwhooker** [comment]
What ever happened to this PR? It looks like ti was ready but seems to have disappeared

**Pull request closed without merging**